### ISSUE TITLE
DateTimeAssertions.BeWithin and BeLessThan do not work if date is on the opposite side of reference date in .Before/.After

### ISSUE THREAD

**Issue by itarob** [fluentassertions/fluentassertions#1467]
I think, `DateTimeAssertions.BeWithin()` and `BeLessThan()` are not working correctly when the validated DateTime is on the opposite side of the reference DateTime and the condition used (`.Before()`/`.After()`)

Instead of trying to textually explain the problem, have a look at the following unit test, which shows the bug for `.BeWithin()` and `.BeLessThan()` (and also shows, `.BeMoreThan()` and `.BeAtLeast()` work as expected)

```cs
[Test]
public void TestFluentAssertion_DateTimeAssertions()
{
	var nowMinus25hrs = DateTime.Now.AddHours(-25);
	var nowMinus23hrs = DateTime.Now.AddHours(-23);
	var nowPlus23hrs = DateTime.Now.AddHours(23);
	var nowPlus25hrs = DateTime.Now.AddHours(25);

	using (new AssertionScope())
	{
		nowMinus25hrs.Should().BeAtLeast(1.Days()).Before(DateTime.Now, "-25, BeAtLeast Before");  // PASS, OK
		nowMinus23hrs.Should().BeAtLeast(1.Days()).Before(DateTime.Now, "-23, BeAtLeast Before");  // FAIL, OK
		nowPlus23hrs.Should().BeAtLeast(1.Days()).Before(DateTime.Now, "23, BeAtLeast Before");  // FAIL, OK
		nowPlus25hrs.Should().BeAtLeast(1.Days()).Before(DateTime.Now, "25, BeAtLeast Before");  // FAIL, OK

		nowMinus25hrs.Should().BeAtLeast(1.Days()).After(DateTime.Now, "-25, BeAtLeast After");  // FAIL, OK
		nowMinus23hrs.Should().BeAtLeast(1.Days()).After(DateTime.Now, "-23, BeAtLeast After");  // FAIL, OK
		nowPlus23hrs.Should().BeAtLeast(1.Days()).After(DateTime.Now, "23, BeAtLeast After");  // FAIL, OK
		nowPlus25hrs.Should().BeAtLeast(1.Days()).After(DateTime.Now, "25, BeAtLeast After");  // PASS, OK

		nowMinus25hrs.Should().BeMoreThan(1.Days()).Before(DateTime.Now, "-25, BeMoreThan Before");  // PASS, OK
		nowMinus23hrs.Should().BeMoreThan(1.Days()).Before(DateTime.Now, "-23, BeMoreThan Before");  // FAIL, OK
		nowPlus23hrs.Should().BeMoreThan(1.Days()).Before(DateTime.Now, "23, BeMoreThan Before");  // FAIL, OK
		nowPlus25hrs.Should().BeMoreThan(1.Days()).Before(DateTime.Now, "25, BeMoreThan Before");  // FAIL, OK

		nowMinus25hrs.Should().BeMoreThan(1.Days()).After(DateTime.Now, "-25, BeMoreThan After");  // FAIL, OK
		nowMinus23hrs.Should().BeMoreThan(1.Days()).After(DateTime.Now, "-23, BeMoreThan After");  // FAIL, OK
		nowPlus23hrs.Should().BeMoreThan(1.Days()).After(DateTime.Now, "23, BeMoreThan After");  // FAIL, OK
		nowPlus25hrs.Should().BeMoreThan(1.Days()).After(DateTime.Now, "25, BeMoreThan After");  // PASS, OK
		
		nowMinus25hrs.Should().BeWithin(1.Days()).Before(DateTime.Now, "-25, BeWithin Before");  // FAIL, OK
		nowMinus23hrs.Should().BeWithin(1.Days()).Before(DateTime.Now, "-23, BeWithin Before");  // PASS, OK
		nowPlus23hrs.Should().BeWithin(1.Days()).Before(DateTime.Now, "23, BeWithin Before");  // SHOULD FAIL, NOT OK!!!
		nowPlus25hrs.Should().BeWithin(1.Days()).Before(DateTime.Now, "25, BeWithin Before");  // SHOULD FAIL, NOT OK!!!

		nowMinus25hrs.Should().BeWithin(1.Days()).After(DateTime.Now, "-25, BeWithin After");  // SHOULD FAIL, NOT OK!!!
		nowMinus23hrs.Should().BeWithin(1.Days()).After(DateTime.Now, "-23, BeWithin After");  // SHOULD FAIL, NOT OK!!!
		nowPlus23hrs.Should().BeWithin(1.Days()).After(DateTime.Now, "23, BeWithin After");  // PASS, OK
		nowPlus25hrs.Should().BeWithin(1.Days()).After(DateTime.Now, "25, BeWithin After");  // FAIL, OK

		nowMinus25hrs.Should().BeLessThan(1.Days()).Before(DateTime.Now, "-25, BeLessThan Before");  // FAIL, OK
		nowMinus23hrs.Should().BeLessThan(1.Days()).Before(DateTime.Now, "-23, BeLessThan Before");  // PASS, OK
		nowPlus23hrs.Should().BeLessThan(1.Days()).Before(DateTime.Now, "23, BeLessThan Before");  // SHOULD FAIL, NOT OK!!!
		nowPlus25hrs.Should().BeLessThan(1.Days()).Before(DateTime.Now, "25, BeLessThan Before");  // SHOULD FAIL, NOT OK!!!

		nowMinus25hrs.Should().BeLessThan(1.Days()).After(DateTime.Now, "-25, BeLessThan After");  // SHOULD FAIL, NOT OK!!!
		nowMinus23hrs.Should().BeLessThan(1.Days()).After(DateTime.Now, "-23, BeLessThan After");  // SHOULD FAIL, NOT OK!!!
		nowPlus23hrs.Should().BeLessThan(1.Days()).After(DateTime.Now, "23, BeLessThan After");  // PASS, OK
		nowPlus25hrs.Should().BeLessThan(1.Days()).After(DateTime.Now, "25, BeLessThan After");  // FAIL, OK
	}
}
```

In my oppinion, the validations above with the comment 'NOT OK' should fail, but they don't.

I'm unsing FluentAssertions 5.10.3 with .NetCore 3.1

**Comment by itarob** [comment]
I think, in DateTimeRangeAssertions, for these two predicates, the lambda is insufficient:

      [TimeSpanCondition.Within] = new TimeSpanPredicate((Func<TimeSpan, TimeSpan, bool>) ((ts1, ts2) => ts1 <= ts2), "within"),
      [TimeSpanCondition.LessThan] = new TimeSpanPredicate((Func<TimeSpan, TimeSpan, bool>) ((ts1, ts2) => ts1 < ts2), "less than")

it should be ((ts1, ts2) => 0 <= ts1 && ts1 <= ts2) and  ((ts1, ts2) => 0 <= ts1 && ts1 < ts2) respectively.

**Comment by jnyrup** [comment]
Hi @itarob 
Thanks for reporting this and for the exhaustive example of test combinations.

We're open for a PR if you would like to fix it.

Note for anyone who fixes this, the bug exists in both `DateTimeRangeAssertions` and `DateTimeOffsetRangeAssertions`.

**Comment by jnyrup** [comment]
Digging more, it seems it has already been fixed by #1313 which is available from 6.0.0 Alpha 1
https://www.nuget.org/packages/FluentAssertions/6.0.0-alpha0001

**Comment by jnyrup** [comment]
Closing as this should be fixed with 6.0.0 alpha1.
